File: cf-hostname-reconciler/src/reconciler/diff.py
"""Bidirectional diff between desired labels and actual Cloudflare custom hostnames."""

from __future__ import annotations

from dataclasses import dataclass, field

from .cloudflare import CustomHostname
from .docker_state import DesiredHostname


@dataclass(frozen=True)
class Plan:
    to_create: list[DesiredHostname] = field(default_factory=list)
    to_delete: list[CustomHostname] = field(default_factory=list)
    unchanged: list[str] = field(default_factory=list)

    @property
    def empty(self) -> bool:
        return not self.to_create and not self.to_delete


def build_plan(
    desired: dict[str, DesiredHostname], actual: list[CustomHostname]
) -> Plan:
    actual_by_host = {entry.hostname: entry for entry in actual}

    to_create = [entry for host, entry in sorted(desired.items()) if host not in actual_by_host]
    to_delete = [
        entry for host, entry in sorted(actual_by_host.items()) if host not in desired
    ]
    unchanged = sorted(set(desired) & set(actual_by_host))
    return Plan(to_create=to_create, to_delete=to_delete, unchanged=unchanged)

File: cf-hostname-reconciler/src/diff.js
// Bidirectional diff between desired labels and actual Cloudflare custom hostnames.

// desired: Map<hostname, {hostname, zoneHint, container}>
// actual:  Array<{id, hostname, zoneId, zoneName, status}>
export function buildPlan(desired, actual) {
  const actualByHost = new Map(actual.map((e) => [e.hostname, e]));

  const toCreate = [...desired.keys()]
    .sort()
    .filter((h) => !actualByHost.has(h))
    .map((h) => desired.get(h));

  const toDelete = [...actualByHost.keys()]
    .sort()
    .filter((h) => !desired.has(h))
    .map((h) => actualByHost.get(h));

  const unchanged = [...desired.keys()].filter((h) => actualByHost.has(h)).sort();

  return { toCreate, toDelete, unchanged, get empty() {
    return toCreate.length === 0 && toDelete.length === 0;
  } };
}
